File: Temp/Java/Practice/src/sort/MergeSort.java
package sort;

public class MergeSort {
	
	int[] arr;
	
	public MergeSort(int[] arr) {
		this.arr = arr;
	}
	
	public void solve() {
		System.out.println("Merge Sort:");
		mergeSort(0, this.arr.length - 1);
	}
	
	void mergeSort( int left, int right) {
		if (left < right) {
			int middle = (right + left) / 2;
			
			mergeSort(left, middle);
			mergeSort(middle + 1, right);
			
			merge(left, middle, right);
		}
	}
	
	void merge(int left, int middle, int right) {
		int lSize = middle +  1 - left;
		int rSize = right - middle;
		
		int[] L = new int[lSize];
		int[] R = new int[rSize];
		
		for (int index = 0; index < lSize; index++)
			L[index] = arr[left + index];
		for (int index = 0; index < rSize; index++)
			R[index] = arr[middle + 1 + index];
		
		int lIndex, rIndex, arrIndex;
		lIndex = rIndex = 0;
		arrIndex = left;
		
		while (lIndex < lSize && rIndex < rSize)
			arr[arrIndex++] = L[lIndex] < R[rIndex] ? L[lIndex++] : R[rIndex++];
			
		while (lIndex < lSize)
			arr[arrIndex++] = L[lIndex++];
		while (rIndex < rSize)
			arr[arrIndex++] = R[rIndex++];
	}
	

}

File: Temp/Python/Practice/Sort/MergeSort.py
class MergeSort(object):
    def __init__(self, arr):
        super(MergeSort, self).__init__()
        self.arr = arr

    def solve(self):
        self.__merge_sort(self.arr)

    def __merge_sort(self, arr):
        if len(arr) > 1:
            middle = len(arr) // 2

            L = arr[:middle]
            R = arr[middle:]

            self.__merge_sort(L)
            self.__merge_sort(R)

            self.__merge(arr, L, R)

    @staticmethod
    def __merge(arr, L, R):
        index_l = index_r = index_arr = 0
        while index_l < len(L) and index_r < len(R):
            if L[index_l] < R[index_r]:
                arr[index_arr] = L[index_l]
                index_l += 1
            else:
                arr[index_arr] = R[index_r]
                index_r += 1
            index_arr += 1

        for kth, ith in zip(range(index_arr, len(L)), range(index_l, len(L))):
            arr[kth] = L[ith]
        for kth, ith in zip(range(index_arr, len(R)), range(index_r, len(R))):
            arr[kth] = R[ith]


def get_instance(arr=[]):
    return MergeSort(arr=arr)
